src/llvm-late-gc-lowering.cpp

```diff
@@ -0,0 +1,1351 @@
+    // in this BB (i.e. even when they don't participate in the dataflow
+    // computation)
+    BBS.UpExposedUses |= BBS.UpExposedUsesUnrooted;
+    BBS.UpExposedUsesUnrooted.reset();
+    S.LiveSets.push_back(BBS.UpExposedUses);
+    S.Rootings.push_back(std::set<int>{});
+    S.LiveIfLiveOut.push_back(std::vector<int>{});
+    return Number;
+}
+
+void LateLowerGCFrame::NoteUse(State &S, BBState &BBS, Value *V, BitVector &Uses) {
+    // Short circuit to have to avoid dealing specially with vectors of
+    // constants, etc.
+    if (isa<Constant>(V))
+        return;
+    else if (isSpecialPtrVec(V->getType())) {
+        std::vector<int> Nums = NumberVector(S, V);
+        for (int Num : Nums) {
+            MaybeResize(BBS, Num);
+            Uses[Num] = 1;
+        }
+    }
+    else {
+        int Num = Number(S, V);
+        if (Num == -1)
+            return;
+        MaybeResize(BBS, Num);
+        Uses[Num] = 1;
+    }
+}
+
+void LateLowerGCFrame::NoteOperandUses(State &S, BBState &BBS, User &UI, BitVector &Uses) {
+    for (Use &U : UI.operands()) {
+        Value *V = U;
+        if (!isSpecialPtr(V->getType()))
+            continue;
+        NoteUse(S, BBS, V, Uses);
+    }
+}
+
+template <typename VisitInst, typename callback>
+void RecursivelyVisit(callback f, Value *V) {
+    for (Use &VU : V->uses()) {
+        User *TheUser = VU.getUser();
+        if (isa<VisitInst>(TheUser))
+            f(VU);
+        if (isa<CallInst>(TheUser) || isa<LoadInst>(TheUser) ||
+            isa<SelectInst>(TheUser) || isa<PHINode>(TheUser) ||
+            isa<StoreInst>(TheUser))
+            continue;
+        if (isa<GetElementPtrInst>(TheUser) || isa<BitCastInst>(TheUser) || isa<AddrSpaceCastInst>(TheUser)) {
+            RecursivelyVisit<VisitInst, callback>(f, TheUser);
+            continue;
+        }
+        V->dump();
+        TheUser->dump();
+        assert(false && "Unexpected instruction");
+    }
+}
+
+template <typename callback>
+void RecursivelyVisitStoresTo(callback f, Value *V) {
+    RecursivelyVisit<StoreInst>(
+      [&](Use &VU) {
+        assert(isa<StoreInst>(VU.getUser()));
+        if (VU.getOperandNo() != 1) {
+            // We spill jlcall arg arrays to the stack for
+            // better debugging. These have no uses, so we're
+            // ok. Allow that as a special case.
+            assert(cast<StoreInst>(VU.getUser())->getPointerOperand()->getNumUses() == 1);
+            return;
+        }
+        f(cast<StoreInst>(VU.getUser()));
+      }, V);
+}
+
+/*
+ * Mark phi uses in jlcall frames. Note this does not do any sort of phi lifting
+ * which is done only if a phi actually ends up getting used somewhere.
+ */
+void LateLowerGCFrame::MaybeNotePhiJLCallFrameUses(State &S, BBState &BBS, PHINode *Phi)
+{
+    if (!Phi->getType()->isPointerTy())
+        return;
+    Type *ElType = cast<PointerType>(Phi->getType())->getElementType();
```

Comment:
Well, it could check the store type for every phi, or it could do something better. I think it's fine for now.